File: Tests/Source/Integration/PushChannelTests.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


@import WireMockTransport;
@import WireDataModel;

#import "ZMUserSession+Internal.h"
#import "ZMMissingUpdateEventsTranscoder+Internal.h"
#import "WireSyncEngine_iOS_Tests-Swift.h"


@interface PushChannelTests : IntegrationTest

@end


@implementation PushChannelTests

- (void)setUp
{
    [super setUp];
    
    [self createSelfUserAndConversation];
    [self createExtraUsersAndConversations];
}

- (void)testThatWeReceiveRemoteMessagesWhenThePushChannelIsUp
{
    // given
    NSString *testMessage1 = [NSString stringWithFormat:@"%@ message 1", self.name];
    NSString *testMessage2 = [NSString stringWithFormat:@"%@ message 22", self.name];
    
    XCTAssertTrue([self login]);
    ZMUser *sender = [self userForMockUser:self.user1];

    // when
    [self.mockTransportSession performRemoteChanges:^(MockTransportSession<MockTransportSessionObjectCreation> *session) {
        NOT_USED(session);
        // send new message remotely
        ZMGenericMessage *message = [ZMGenericMessage messageWithContent:[ZMText textWith:testMessage1 mentions:@[] linkPreviews:@[] replyingTo:nil] nonce:NSUUID.createUUID];
        MockUserClient *fromClient = self.user1.clients.anyObject, *toClient = self.selfUser.clients.anyObject;
        [self.groupConversation encryptAndInsertDataFromClient:fromClient toClient:toClient data:message.data];
        [self spinMainQueueWithTimeout:0.2];
        
        ZMGenericMessage *secondMessage = [ZMGenericMessage messageWithContent:[ZMText textWith:testMessage2 mentions:@[] linkPreviews:@[] replyingTo:nil] nonce:NSUUID.createUUID];
        [self.groupConversation encryptAndInsertDataFromClient:fromClient toClient:toClient data:secondMessage.data];
    }];
    WaitForAllGroupsToBeEmpty(0.5);
    WaitForAllGroupsToBeEmpty(0.5);

    // then
    ZMConversation *groupConversation = [self conversationForMockConversation:self.groupConversation];
    id<ZMConversationMessage> message1 = [groupConversation lastMessagesWithLimit:50][1];
    id<ZMConversationMessage> message2 = [groupConversation lastMessagesWithLimit:50][0];
    
    XCTAssertEqualObjects(message1.textMessageData.messageText, testMessage1);
    XCTAssertEqualObjects(message2.textMessageData.messageText, testMessage2);
    XCTAssertEqualObjects(message1.sender, sender);
    XCTAssertEqualObjects(message2.sender, sender);
}

- (void)testThatItFetchesLastNotificationsFromBackendIgnoringTransientNotificationsID
{
    // given
    XCTAssertTrue([self login]);
    
    [self.mockTransportSession performRemoteChanges:^(MockTransportSession<MockTransportSessionObjectCreation> *session) {
        NOT_USED(session);
        // will create a transient notification
        [session sendIsTypingEventForConversation:self.groupConversation user:self.user1 started:YES];
    }];
    WaitForAllGroupsToBeEmpty(0.5);
    
    [self.mockTransportSession performRemoteChanges:^(MockTransportSession<MockTransportSessionObjectCreation> *session) {
            NOT_USED(session);
        // will create a notification that is not transient
        ZMGenericMessage *message = [ZMGenericMessage messageWithContent:[ZMText textWith:@"Food" mentions:@[] linkPreviews:@[] replyingTo:nil] nonce:NSUUID.createUUID];
        [self.groupConversation encryptAndInsertDataFromClient:self.user1.clients.anyObject toClient:self.selfUser.clients.anyObject data:message.data];
    }];
    WaitForAllGroupsToBeEmpty(0.5);
    
    __block NSUUID *messageAddLastNotificationID;
    [self.mockTransportSession performRemoteChanges:^(MockTransportSession<MockTransportSessionObjectCreation> *session) {
        NOT_USED(session);
        // save previous notification ID
        MockPushEvent *messageEvent = self.mockTransportSession.updateEvents.lastObject;
        messageAddLastNotificationID = messageEvent.uuid;
        XCTAssertEqualObjects(messageEvent.payload.asDictionary[@"type"], @"conversation.otr-message-add");
        
        // will create a transient notification
        [session sendIsTypingEventForConversation:self.groupConversation user:self.user1 started:NO];
        
    }];
    WaitForAllGroupsToBeEmpty(0.5);
    [self.mockTransportSession resetReceivedRequests];
    
    // when
    [self.mockTransportSession performRemoteChanges:^(MockTransportSession<MockTransportSessionObjectCreation> *session) {
        [session simulatePushChannelClosed];
        [session simulatePushChannelOpened];
    }];
    WaitForAllGroupsToBeEmpty(0.5);

    // then
    NSString *expectedLastRequest = [NSString stringWithFormat:@"/notifications?size=%lu&since=%@&client=%@", ZMMissingUpdateEventsTranscoderListPageSize, messageAddLastNotificationID.transportString, self.userSession.selfUserClient.remoteIdentifier];
    XCTAssertEqualObjects([(ZMTransportRequest *)self.mockTransportSession.receivedRequests.lastObject path], expectedLastRequest);
}



@end

File: Tests/Source/Integration/PushChannelTests.swift
//
// Wire
// Copyright (C) 2020 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation
import WireDataModel
import WireMockTransport
@testable import WireSyncEngine

class PushChannelTests: IntegrationTest {
    
    override func setUp() {
        super.setUp()
        self.createSelfUserAndConversation()
        self.createExtraUsersAndConversations()
    }
    
    func testThatWeReceiveRemoteMessagesWhenThePushChannelIsUp() {
        // GIVEN
        let testMessage1 = "\(name) message 1"
        let testMessage2 = "\(name) message 2"
        
        XCTAssertTrue(login())
        let sender = user(for: user1)
        
        // WHEN
        mockTransportSession.performRemoteChanges { _ in
            let message = GenericMessage(content: Text(content: testMessage1), nonce: .create())
           
            guard
                let fromClient = self.user1.clients.anyObject() as? MockUserClient,
                let toClient = self.selfUser.clients.anyObject() as? MockUserClient,
                let data1 = try? message.serializedData() else {
                    return XCTFail()
            }
            self.groupConversation.encryptAndInsertData(from: fromClient, to: toClient, data: data1)
            self.spinMainQueue(withTimeout: 0.2)
            
            let secondMessage = GenericMessage(content: Text(content: testMessage2), nonce: .create())
            guard let data2 = try? secondMessage.serializedData() else {
                return XCTFail()
            }
            self.groupConversation.encryptAndInsertData(from: fromClient, to: toClient, data: data2)
        }
        _ = waitForAllGroupsToBeEmpty(withTimeout: 0.5)
        
        // THEN
        let conversation = self.conversation(for: groupConversation)
        let message1 = conversation?.lastMessages(limit: 50)[1]
        let message2 = conversation?.lastMessages(limit: 50)[0]
        
        XCTAssertEqual(message1?.textMessageData?.messageText, testMessage1)
        XCTAssertEqual(message2?.textMessageData?.messageText, testMessage2)
        XCTAssertEqual(message1?.sender, sender)
        XCTAssertEqual(message2?.sender, sender)
    }
    
    func testThatItFetchesLastNotificationsFromBackendIgnoringTransientNotificationsID() {
        // GIVEN
        XCTAssertTrue(login())
        
        mockTransportSession.performRemoteChanges { _ in
            // will create a transient notification
            self.mockTransportSession.sendIsTypingEvent(for: self.groupConversation, user: self.user1, started: true)
        }
        _ = waitForAllGroupsToBeEmpty(withTimeout: 0.5)
        
        mockTransportSession.performRemoteChanges { _ in
            // will create a notification that is not transient
            let message = GenericMessage(content: Text(content: "Foo"), nonce: .create())

            guard
                let fromClient = self.user1.clients.anyObject() as? MockUserClient,
                let toClient = self.selfUser.clients.anyObject() as? MockUserClient,
                let data = try? message.serializedData() else {
                    return XCTFail()
            }
            
            self.groupConversation.encryptAndInsertData(from: fromClient, to: toClient, data: data)
        }
        _ = waitForAllGroupsToBeEmpty(withTimeout: 0.5)
        
        var messageAddLastNotificationID: UUID?
        mockTransportSession.performRemoteChanges { _ in
            // save previous notification ID
            let messageEvent = self.mockTransportSession.updateEvents.last as? MockPushEvent
            messageAddLastNotificationID = messageEvent?.uuid
            XCTAssertEqual(messageEvent?.payload.asDictionary()?["type"] as? String, "conversation.otr-message-add")
            
            // will create a transient notificaiton
            self.mockTransportSession.sendIsTypingEvent(for: self.groupConversation, user: self.user1, started: false)
        }
        _ = waitForAllGroupsToBeEmpty(withTimeout: 0.5)
        
        mockTransportSession.resetReceivedRequests()
        
        // WHEN
        mockTransportSession.performRemoteChanges { session in
            session.simulatePushChannelClosed()
            session.simulatePushChannelOpened()
        }
        _ = waitForAllGroupsToBeEmpty(withTimeout: 0.5)
        
        // THEN
        let expectedLastRequest = "/notifications?size=\(ZMMissingUpdateEventsTranscoderListPageSize)&since=\(messageAddLastNotificationID?.transportString() ?? "")&client=\(userSession?.selfUserClient?.remoteIdentifier ?? "")"
        XCTAssertEqual(mockTransportSession.receivedRequests().last?.path , expectedLastRequest)
    }
}
